### PR TITLE
Shields Re-Rework

### PR DESCRIPTION
**Pull request by SethDGamre**
Downtime is dead. Long live overkill cost.

otherwise implemented in accordance with the shield rework outlined in the game design document. 

The last shot that kills the shield  will keep it deactivated for as long as it takes for the regenspeed and regencost to pay for the overkill damage and come back online. This should mean that it should plug and play with the existing game balance.

plasma (cannon, EMG) projectiles are the only things blocked by shields in the rework.

un-hid the modoption.

testing steps: enable "shields re-rework v2.0" in the experimental modoptions and test various cannon and emg style projectiles on the shields. Should be compatible with the other shield behavior overrides (except it'll overwrite the "bounce" options) and be compatible with the evocom cortex shields (until I rework cortex shields)

TL;DR new features:
- shields only block plasma
- improved shield protection - you shouldn't be able to damage units with AOE  through a shield even right up to the edge
- you can turn shields on and off
- when a shield is hit with damage > remaining charge, it'll take "overkill damage", it must recharge the overkill damage before coming back online at normal recharge cost
- while recovering from overkill damage, shield turns off

### CODE REVIEW

File: luarules/gadgets/lups_shield.lua [hunk 1]
```diff
@@ -23,6 +23,7 @@ end
 local GAMESPEED = Game.gameSpeed
 local SHIELDARMORID = 4
 local SHIELDARMORIDALT = 0
+local SHIELDONRULESPARAMINDEX = 531313
```
Comment (sprunk) [inline luarules/gadgets/lups_shield.lua:26]: Rules params accept strings as keys, why not use those?

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 2]
```diff
@@ -182,6 +246,22 @@ local function shieldNegatesDamageCheck(unitID, unitTeam, attackerID, attackerTe
 	return false
 end
 
+local function setProjectilesAlreadyInsideShield(shieldUnitID, radius)
+	-- This section is to allow slower moving projectiles already inside the shield when it comes back online to damage units within the radius.
+	local x, y, z = spGetUnitPosition(shieldUnitID)
+	-- Engine has GetProjectilesInRectangle, but not GetProjectilesInCircle, so we have to square the circle
+	-- TODO: Change to GetProjectilesInCircle once it is added
+	local radius = radius * math.sqrt(math.pi) / 2
```
Comment (sprunk) [inline luarules/gadgets/unit_shield_behaviour.lua:254]: Looks suspicious, why sqrt(pi)? I know it's just copying existing (and presumably working good enough) code but perhaps it's an opportunity to look at.

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 3]
```diff
@@ -158,8 +222,8 @@ local function triggerDowntime(unitID, weaponNum)
 	spSetUnitShieldRechargeDelay(unitID, weaponNum, 3600)
 
 	spSetUnitShieldState(unitID, weaponNum, false)
-	shieldData.downtimeReset = gameSeconds + shieldData.downtime
 	shieldData.shieldEnabled = false
+	spSetUnitRulesParam(unitID, shieldOnUnitRulesParamIndex, 0, {inlos = true})
```
Comment (sprunk) [inline luarules/gadgets/unit_shield_behaviour.lua:226]: Perhaps do something like `local INLOS = {inlos = true}` at file scope so as not to create garbage tables each time.

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 4]
```diff
@@ -49,29 +49,83 @@ local projectileDefIDCache = {}
...
+	"LaserCannon",
+	"LightningCannon"
+   }
+   local hasDamageFalloff = false
+   for _, exemption in ipairs(damageFalloffUnitTypes) do
+		if exemption == weaponDef.type then
+			hasDamageFalloff = true
+			break
+		end
+	end
```
Comment (sprunk) [inline luarules/gadgets/unit_shield_behaviour.lua:106]: Don't need to iterate manually:
```suggestion
	local damageFalloffUnitTypes = {
		BeamLaser = true,
		Flame = true,
		LaserCannon = true,
		LightningCannon = true,
	}
	if damageFalloffUnitTypes[weaponDef.type] then
		hasDamageFalloff = true
	end
```

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 5]
```diff
@@ -49,29 +49,83 @@ local projectileDefIDCache = {}
...
+	if weaponDef.beamtime and weaponDef.beamtime < 1 then
+		local minimumMinIntensity = 0.5
+		local minIntensity = weaponDef.minIntensity or minimumMinIntensity
+		minIntensity = math.max(minIntensity, minimumMinIntensity)
+		-- This splits up the damage of hitscan weapons over the duration of beamtime, as each frame counts as a hit in ShieldPreDamaged() callin
+		-- Math.floor is used to sheer off the extra digits of the number of frames that the hits occur
+		beamtimeReductionMultiplier = 1 / math.floor(weaponDef.beamtime * Game.gameSpeed)
+   end
+
+   local minimumMinIntensity = 0.65
```
Comment (sprunk) [inline luarules/gadgets/unit_shield_behaviour.lua:93]: Why 0.65?

File: luarules/gadgets/lups_shield.lua (hunk 1 above)
Comment (SethDGamre) [reply luarules/gadgets/lups_shield.lua:26]: Storing and referencing a number key will be more performant than a string if it's called repeatedly like it is here, correct? In which case I would prefer the more performant option as it appears a great deal of effort has gone into making this performant

Comment (sprunk) [reply luarules/gadgets/lups_shield.lua:26]: I expect it's a bit more performant but I don't think it has ever been measured so it's hard to tell by how much. I don't know what the guideline on perf vs maintainabiltity is so it's up to @WatchTheFort. If he agrees to keep it as a number then perhaps document it though:
```suggestion
local SHIELDONRULESPARAMINDEX = 531313 -- not a string due to perfmaxxing
```

Comment (WatchTheFort) [reply luarules/gadgets/lups_shield.lua:26]: How on earth would referencing a number vs a string make any real, measurable performance difference?

Always favour code readability, other devs needs to be able to divine what everything means. What does this variable refer to?

Comment (SethDGamre) [reply luarules/gadgets/lups_shield.lua:26]: the index of a customrulesparam that's checked every update to determine if the shield bubble should be drawn or not.

Comment (SethDGamre) [reply luarules/gadgets/lups_shield.lua:26]: this appears both in the unit_shield_behavior.lua gadget and in the shield_lups.lua gadget, which are the majority of places shield code is handled game-side

File: luarules/gadgets/unit_shield_behaviour.lua (hunk 2 above)
Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua:256]: gonna be honest, I'm not exactly sure what's going on here. Watch the Fort squared my circle for me when the shield turns back on and there's projectiles within its radius.  Copilot seems to think it works great, and it works in game. But copilot tends to think almost anything complicated is great.

File: luarules/gadgets/unit_shield_behaviour.lua (hunk 3 above)
Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua:226]: donezo

File: luarules/gadgets/unit_shield_behaviour.lua (hunk 4 above)
Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua]: yeah, that's way cleaner.

File: luarules/gadgets/unit_shield_behaviour.lua (hunk 5 above)
Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua]: Imperically tested. Added comment. Sufficient?

File: luarules/gadgets/unit_shield_behaviour.lua (hunk 2 above)
Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua:254]: @WatchTheFort

**Comment by Damgam** [comment]
fuck i've merged the wrong pull request
